osquery/tables/events/linux/process_events.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/algorithm/hex.hpp>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+
+#include "osquery/events/linux/audit.h"
+
+namespace osquery {
+
+#define AUDIT_SYSCALL_EXECVE 59
+
+// Depend on the external getUptime table method.
+namespace tables {
+extern long getUptime();
+}
+
+/**
+ * @brief Audit will generate consecutive messages related to a process event.
+ *
+ * The first are the syscall details, which contain information about the
+ * process generating the event. That is followed by the exec arguments
+ * including the canonicalized/interpreted program name. The working directory
+ * and a null-delimited set of path messages follow and complete the set of
+ * information.
+ */
+enum AuditProcessEventState {
+  STATE_SYSCALL = AUDIT_SYSCALL,
+  STATE_EXECVE = AUDIT_EXECVE,
+  STATE_CWD = AUDIT_CWD,
+  STATE_PATH = AUDIT_PATH,
+};
+
+class ProcessEventSubscriber : public EventSubscriber<AuditEventPublisher> {
+ public:
+  /// The process event subscriber declares an audit event type subscription.
+  Status init();
+
+  /// Kernel events matching the event type will fire.
+  Status Callback(const AuditEventContextRef& ec, const void* user_data);
+
+ private:
+  /// The next expected process event state.
+  AuditProcessEventState state_{STATE_SYSCALL};
+
+  /// Along with the state, track the building row.
+  Row row_;
+};
+
+static inline std::string decodeValue(const std::string& s) {
+  return (s[0] == '"') ? s.substr(1, s.size() - 2) : boost::algorithm::unhex(s);
+}
+
+REGISTER(ProcessEventSubscriber, "event_subscriber", "process_events");
+
+Status ProcessEventSubscriber::init() {
+  auto sc = createSubscriptionContext();
+
+  // Monitor for execve syscalls.
+  sc->rules.push_back({59, ""});
+
+  // Request call backs for all parts of the process execution state.
+  // Drop events if they are encountered outside of the expected state.
+  sc->types = {AUDIT_SYSCALL, AUDIT_EXECVE, AUDIT_CWD, AUDIT_PATH};
+  subscribe(&ProcessEventSubscriber::Callback, sc, nullptr);
+
+  return Status(0, "OK");
+}
+
+Status ProcessEventSubscriber::Callback(const AuditEventContextRef& ec,
+                                        const void* user_data) {
+  // Define some acceptable transitions outside of the default state.
```

Comment:
This method should be broken out into a testable handler, OR, subsequent test runs could inspect the RocksDB store for verification. An inline method returning status and NOT reseting the manipulated `row_` data would be ideal.